### PR TITLE
Add editorconfig parser and unit tests

### PR DESCRIPTION
**Pull request by agocke**
This is the same parser used by Visual Studio for editorconfig with a
minor modification to use case-insensitive comparison for property keys.

Right now the parser is unused, but unit tests have been added in preparation for later use.

### CODE REVIEW

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                // which is the recommended Unicode comparison for case-insensitive identifiers. The
+                // implementation performs the equivalent of a lower-case comparsion. The
+                // editorconfig file is defined as using a UTF-8 encoding, making Unicode comparison
+                // relevant.
+                this.Properties = new Dictionary<string, string>(CaseInsensitiveComparison.Comparer);
+            }
+
+            public string Name { get; }
+
+            public IDictionary<string, string> Properties { get; }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: @agocke Just an advisement about the conversation going on over here: https://github.com/dotnet/roslyn/issues/23391

Specifically, we've encountered some undefined behavior issues because some Roslyn features want to know the specific order of items in .editor_config.  This is because some features want to have a "first applicable editor_config entry 'wins'" strategy.  If we are to supply editorconfig support at the CodeAnalysis layer, it would be nice to be able to provide that.

I'll loop you into that issue.

File: src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+using System.Linq;
+using Roslyn.Test.Utilities;
+using Roslyn.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.UnitTests
+{
+    public class EditorConfigTests
+    {
+        private static EditorConfig ParseConfigFile(string text) => EditorConfig.Parse(text, "");
```
Comment (jaredpar) [inline src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs]: >private static EditorConfig ParseConfigFile(string text) => EditorConfig.Parse(text, ""); [](start = 8, length = 89)

Should this use the ` {get; } = ` style to avoid re-parsing an empty string on every call?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+using System.Collections.Generic;
+using System.IO;
+using System.Text.RegularExpressions;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a single EditorConfig file, see http://editorconfig.org for details about the format.
+    /// </summary>
+    internal class EditorConfig
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: sealed

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+    {
+        // Matches EditorConfig section header such as "[*.{js,py}]", see http://editorconfig.org for details
+        private static Regex _sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        // Matches EditorConfig property such as "indent_style = space", see http://editorconfig.org for details
+        private static Regex _propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        public Section GlobalSection = new Section(string.Empty);
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Convert to a property

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        public string Directory { get; private set; }
+
+        public IList<Section> NamedSections = new List<Section>();
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Convert to a property

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+        }
+
+        /// <summary>
+        /// Gets whether this editorconfig is a the topmost editorconfig.
+        /// </summary>
+        public bool IsRoot
+        {
+            get
+            {
+                return this.GlobalSection.Properties.ContainsKey("root") && this.GlobalSection.Properties["root"] == "true";
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: >e"; [](start = 121, length = 3)

is `"true"` case sensitive in editro config?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+
+            return editorConfig;
+        }
+
+        private static bool IsComment(string line)
+        {
+            for (int i = 0; i < line.Length; i++)
+            {
+                char c = line[i];
+                if (!Char.IsWhiteSpace(c))
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Use `char` not `Char`

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+                {
+                    return c == '#' || c == ';';
+                }
+            }
+
+            return false;
+        }
+
+        internal class Section
```
Comment (jaredpar) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: sealed

**Comment by jaredpar** [comment]
Done with review pass 1

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 1 above)
Comment (tannergooding) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: > This is because some features want to have a "first applicable editor_config entry 'wins'" strategy. 

Naively, that seems like the opposite of what users will expect. EditorConfig files are imported and parsed from root downward (to the current directory). The standard behavior according to the editorconfig "spec" is last definition wins.

> When opening a file, EditorConfig plugins look for a file named .editorconfig in the directory of the opened file and in every parent directory. A search for .editorconfig files will stop if the root filepath is reached or an EditorConfig file with root=true is found.

> EditorConfig files are read top to bottom and the closest EditorConfig files are read last. Properties from matching EditorConfig sections are applied in the order they were read, so properties in closer files take precedence.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 6 above)
Comment (tannergooding) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: > The section names are filepath globs (case sensitive), similar to the format accepted by gitignore.

> Currently all properties and values are case-insensitive. They are lowercased when parsed.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Let me rephrase:  the IDE wants to use order to define some behavioral contract.  Whether that is "first wins" or "last wins" and how that interacts with rules that are specified from outside-in, still needs to be committed to.  If you want, you can def provide feedback on the linked bug where this is being discussed.  The critical bit for IDE is that the API for obtaining this information be rich enough to at least supply whatever contract is decided upon, without depending on undefined behavior.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+            GlobalSection = globalSection;
+            NamedSections = namedSections;
+            Directory = directory;
+        }
+
+        /// <summary>
+        /// Gets whether this editorconfig is a the topmost editorconfig.
+        /// </summary>
+        public bool IsRoot
+            => GlobalSection.Properties.TryGetValue("root", out string val) && val == "true";
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: is this intentionally case sensitive?  if so, can you document as such?

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Are values in the properties bag trimmed?  i.e. if i have ```root = true ``` could that be a problem?  (perhaps good to test).

File: src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+            var config = ParseConfigFile(@"
+long: this value continues
+   in the next line");
+
+            var properties = config.GlobalSection.Properties;
+            Assert.Equal(1, properties.Count);
+            AssertEx.SetEqual(
+                new[] { KeyValuePair.Create("long", "this value continues")},
+                properties);
+        }
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/CodeAnalysisTest/EditorConfigTests.cs]: expected a test of "Root"

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 9 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: It's not case sensitive. Note that we force a ToLower on keys and values now.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+
+        /// <summary>
+        /// Parses an editor config file text located within the given parent directory.
+        /// </summary>
+        public static EditorConfig Parse(string text, string parentDirectory)
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: can this document the behavior that will happen on errors?  will this always succeed?  can there be parsing errors?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+                        {
+                            // This is the global section
+                            globalSection = section;
+                        }
+                        else
+                        {
+                            namedSectionBuilder.Add(section);
+                        }
+
+                        var sectionName = sectionMatches[0].Groups[1].Value;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: note: regexes support named groups.  That may help make this code clearer.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 13]
```diff
@@ -0,0 +1,156 @@
...
+                        activeSectionProperties = ImmutableDictionary.CreateBuilder<string, string>(
+                            CaseInsensitiveComparison.Comparer);
+                        continue;
+                    }
+
+                    var propMatches = _propertyMatcher.Matches(line);
+                    if (propMatches.Count > 0 && propMatches[0].Groups.Count > 1)
+                    {
+                        var key = propMatches[0].Groups[1].Value;
+                        var value = propMatches[0].Groups[2].Value;
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: same feedback about named groups here.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 14]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a single EditorConfig file, see http://editorconfig.org for details about the format.
+    /// </summary>
+    internal sealed class EditorConfig
+    {
+        // Matches EditorConfig section header such as "[*.{js,py}]", see http://editorconfig.org for details
+        private static readonly Regex _sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: named groups

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 11 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: oddly, this is called "parentdirectory" but is called "directory" everywhere else.  intentional?

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 15]
```diff
@@ -0,0 +1,156 @@
...
+        public bool IsRoot
+            => GlobalSection.Properties.TryGetValue("root", out string val) && val == "true";
+
+        /// <summary>
+        /// Parses an editor config file text located within the given parent directory.
+        /// </summary>
+        public static EditorConfig Parse(string text, string parentDirectory)
+        {
+            Section globalSection = null;
+            var namedSectionBuilder = ImmutableArray.CreateBuilder<Section>();
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: ArrayBuilder+ToImmutableAndFree?  Otherwise, this will generate garbage.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 9 above)
Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Ah, good to know.  Probably good for the docs on the Properties dict exposed off of sections.  (or, if it's already there, my bad that i missed it).

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 15 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: If this code is shared between multiple codebases we can't use Roslyn internal types. I was trying to avoid it for now.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+            // N.B. The editorconfig documentation is quite loose on property interpretation.
+            // Specifically, it says:
+            //      Currently all properties and values are case-insensitive.
+            //      They are lowercased when parsed.
+            // To accomodate this, we use a lower case Unicode mapping when adding to the
+            // dictionary, but we also use a case-insensitve key comparer when doing lookups
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: ❓ What will happen to this?

https://github.com/pkzxs/Aurora.Music/blob/8b6a97f9d8457b834dc86eea573c75410ea8dea7/Source/.editorconfig#L18

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: It will be lowercased, as required by the spec.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: That will break existing roslyn IDE scenarios that customers are already depending on... Can we not lowercase and allow clients of the parser to decide the semantics they want for this sort of thing?

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: If we end up using the case sensitive version there was no point in lowercasing it anyway. If we want to be non-compliant we should do so strongly and clearly.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: > If we want to be non-compliant we should do so strongly and clearly.

That was my intent when i had CaseSensitiveProperties and CaseInsensitiveProperties.  It makes it very clear what you're getting, and anyone consuming this has to consider which is the right thing for their domain.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Note: i can see info about things being case insensitive here: https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties

But it makes it sound more like normative text letting you know that for those specific properties, things are case insensitive, not necessarily a blanket rule that all property values are case insensitive.

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: @sharwell So the parser should recognize the properties listed on the website and lowercase them, but everything else is our decision? And any new properties get added to the special list?

@CyrusNajmabadi You also intend this to apply to keys? Or just values?

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: This link (http://editorconfig.org/#file-format-details) happens to mention: ```Currently all properties and values are case-insensitive. They are lowercased when parsed.```

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: @agocke That's what this library does:

https://github.com/editorconfig/editorconfig-core-net/blob/master/src/EditorConfig.Core/FileConfiguration.cs#L95

lowercases the specified ones, does not touch the rest.  That's the same with this js impl: https://github.com/editorconfig/editorconfig-core-js/blob/master/src/index.ts#L125

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Note: thse were two actual impls linked to from the main editorconfig page.  So i thought i would take a look to see what the rest of the ecosystem is doing.

Comment (sharwell) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: @CyrusNajmabadi Also interesting to note that keys are always converted to lowercase. Only the values get special handling depending on the key name.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: C version also only lowercases the values of the well known set: https://github.com/editorconfig/editorconfig-core-c/blob/master/src/lib/editorconfig.c#L84

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: Given that this will ship in the compiler, what kind of update contract should apply?

Comment (sharwell) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: > Given that this will ship in the compiler, what kind of update contract should apply?

Proposal:

* Keys are case insensitive and converted to lowercase when parsed
* Values are generally case sensitive, and trimmed but not otherwise modified during parsing
* Values which are defined by editorconfig.org are converted to lowercase if specified by that site, except for cases where the key contains a `.` (no such key currently falls into this exception, but it allows for changes to be made if the core specification changes while still allowing users to define stable property behavior via "namespacing")

Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: That's not an update contract. What happens when the website is updated? Do we promise to update our list? In what timeframe?

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: It woudl be interesting if they updated in a substantial manner.  My personal preference would be that if such a change would cause a break in our existing behavior (i.e. existing keys/values would be interpretted differently), then we would have ways to ask for the specific formats.  

i.e. you could ask for the V1 EditorConfig interpretation (which we would always preserve), or you could ask for the V2 if that comes about.  etc. etc.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 17]
```diff
@@ -0,0 +1,170 @@
...
+{
+    /// <summary>
+    /// Represents a single EditorConfig file, see http://editorconfig.org for details about the format.
+    /// </summary>
+    internal sealed class EditorConfig
+    {
+        // Matches EditorConfig section header such as "[*.{js,py}]", see http://editorconfig.org for details
+        private static readonly Regex _sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        // Matches EditorConfig property such as "indent_style = space", see http://editorconfig.org for details
+        private static readonly Regex _propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: > `([#;].*)`

:bulb: Use `(?:[#;].*)` for efficiency

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 16 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: I've implemented a system of "reserved" keys and values similar to the other implementations. Right now I provide no update contract, only saying that things may be added as we go on. Also, there is no special handling of "." as I didn't think that additional level of complexity was merited.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 18]
```diff
@@ -161,8 +225,11 @@ public Section(string name, ImmutableDictionary<string, string> properties)
             public string Name { get; }
 
             /// <summary>
-            /// Keys and values for this section. All keys and values are lower-cased according to the
-            /// EditorConfig specification and keys are compared case-insensitively.
+            /// Keys and values for this section. All keys are lower-cased according to the
+            /// EditorConfig specification and keys are compared case-insensitively. Values are
+            /// lower-cased if the value appears in <see href="EditorConfig.ReservedValues" />
+            /// or if the corresponding key is in <see href="EditorConfig.ReservedKeys" />. Otherwise,
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: ❕ Both of these should be `cref` instead of `href`

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 19]
```diff
@@ -15,9 +16,69 @@ namespace Microsoft.CodeAnalysis
     internal sealed class EditorConfig
     {
         // Matches EditorConfig section header such as "[*.{js,py}]", see http://editorconfig.org for details
-        private static readonly Regex _sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        private static readonly Regex s_sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
         // Matches EditorConfig property such as "indent_style = space", see http://editorconfig.org for details
-        private static readonly Regex _propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+        private static readonly Regex s_propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        private static ImmutableHashSet<string> s_lazyReservedKeys;
```
Comment (sharwell) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: ❓ Why lazy? Seems like more work and complexity for no real gain.

**Comment by sharwell** [comment]
Also did you see this comment? https://github.com/dotnet/roslyn/pull/26253#discussion_r183182741

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+                        "indent_size",
+                        "tab_width",
+                        "end_of_line",
+                        "charset",
+                        "trim_trailing_whitespace",
+                        "insert_final_newline",
+                    };
+                    Interlocked.CompareExchange(
+                        ref s_lazyReservedKeys,
+                        ImmutableHashSet.CreateRange(CaseInsensitiveComparison.Comparer, reservedKeys),
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: I seems the spec/wiki doesn't say that keys are case-insensitive.

**Comment by agocke** [comment]
@sharwell I'm not interested in messing with perf right now. Let's get it working, then optimize it. Also, we may not be able to ship the regex anyway, so if I'm rewriting it I don't think it's worth spending time on.

**Comment by agocke** [comment]
@jcouv Did you have any more comments?

**Comment by jcouv** [comment]
@agocke I'll look this afternoon.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 21]
```diff
@@ -0,0 +1,206 @@
...
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// Represents a single EditorConfig file, see http://editorconfig.org for details about the format.
+    /// </summary>
+    internal sealed class EditorConfig
+    {
+        // Matches EditorConfig section header such as "[*.{js,py}]", see http://editorconfig.org for details
+        private static readonly Regex s_sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: >s_sectionMatcher [](start = 38, length = 16)

Consider adding some examples and documentation: I see that `#` is the beginning of comments. It would be good to indicate that `;` also mean comments.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 22]
```diff
@@ -0,0 +1,206 @@
...
+        // Matches EditorConfig property such as "indent_style = space", see http://editorconfig.org for details
+        private static readonly Regex s_propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: >(.*?) [](start = 96, length = 5)

`(.*?)`
I didn't understand this. I would have expected `(([^#;]|\\#|\\;)+)` again, optionally.

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs [hunk 23]
```diff
@@ -0,0 +1,206 @@
...
+                "charset",
+                "trim_trailing_whitespace",
+                "insert_final_newline",
+            });
+
+        /// <summary>
+        /// A set of values that are reserved for special use for the editorconfig specification
+        /// and will always be lower-cased by the parser.
+        /// </summary>
+        public static ImmutableHashSet<string> ReservedValues { get; }
```
Comment (jcouv) [inline src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: >{ get; } [](start = 62, length = 8)

Nit: `=>`

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 15 above)
Comment (jcouv) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: It seems this was not addressed yet (`namedSectionBuilder` should be freed).

---
In reply to: [182908455](https://github.com/dotnet/roslyn/pull/26253#discussion_r182908455) [](ancestors = 182908455)

File: src/Compilers/Core/Portable/CommandLine/EditorConfig.cs (hunk 23 above)
Comment (agocke) [reply src/Compilers/Core/Portable/CommandLine/EditorConfig.cs]: That would create a new set of ReservedValues on every call. This is just a readonly property.

**Comment by agocke** [comment]
@jcouv I've taken your comments as a follow-up. https://github.com/dotnet/roslyn/projects/18#card-9700989 Right now I'd prefer to prevent deviance from the VS parser, but I agree that it should be documented and spec'd before shipping.

**Comment by agocke** [comment]
@jaredpar I think I've addressed your comments. Was there anything I missed?
